File: train/train.lua
require 'xlua'

opt = lapp[[
    --seed                  (default 251)                rand seed

    --model                 (default '')                 model name
    --model_param_load      (default '')                 weight file to load
    --bn_meanstd_load       (default '')                 batch normalization file to load
    --optimstate_load       (default '')                 state of optim.adam to load

    --num_frames            (default 5)                  number of frames in input stack
    --num_channels          (default 3)                  rgb input
    --max_intensity         (default 255)                maximum intensity of input

    --data_root             (default '')                 folder for datasets
    --data_trainset         (default '')                 folder for transet data
    --data_validset         (default '')                 folder for validset data
    --trainset_size         (default 61)                 size of trainset (if is larger than dataset, use all dataset as trainset)
    --batch_size            (default 64)                 size of batch sampled from batch files (if is larger than size from batch files, use entire batch every iteration)
    --it_max                (default 80000)              max number of iterations

    --model_param           (default '')                 weight file
    --bn_meanstd            (default '')                 batch normalization file
    --optimstate            (default '')                 state of optim.adam
    --save_every            (default 500)                auto save every save_every iterations
    --log                   (default '')                 dir to save log
    --log_every             (default 1)                  log every log_every iterations

    --reset_lr              (default 0)                  reset lr to reset_lr instead of default or saved lr
    --reset_state           (default 0)                  reset optim state
    --decay_from            (default 24000)              decay learning rate from decay_from iterations
    --decay_every           (default 8000)               decay learning rate every decay_every iteration
    --decay_rate            (default 0.5)                decay learning rate

    --overfit_batches       (default 0)                  overfit batch num for debuging
    --overfit_patches       (default 0)                  overfit patch num, valid if overfit_batches = 1
    --overfit_out           (default '')                 overfit output on trainset
]]
    

-- scan batches
require("lfs")
trainsetDirs = {}
validsetDirs = {}
if #opt.data_root > 0 then

    alignFolder = opt.data_root
    print(string.format('Scanning videos from %s',alignFolder))
    videoNames = {}
    videoName2batchDirs = {}
    local iVideos = 0;
    local nBatches = 0;
    for videoName in lfs.dir(alignFolder) do
        if videoName ~= "." and videoName ~= ".." then
    --         print(videoName)
            iVideos = iVideos+1
            videoNames[iVideos] = videoName
            videoName2batchDirs[videoName] = {};
            local videoFolder = paths.concat(alignFolder,videoName);
            for batchName in lfs.dir(videoFolder) do
                if batchName ~= "." and batchName ~= ".." then
                    table.insert(videoName2batchDirs[videoName],
                        paths.concat(videoFolder,batchName))
    --                 print(batchName)
                end
            end
            table.sort(videoName2batchDirs[videoName])
            nBatches = nBatches + #videoName2batchDirs[videoName]
        end
    end
    table.sort(videoNames)
    print(string.format('Found %d videos and %d batches',#videoNames,nBatches))

    -- split datasets (videoNames) into trainset and validset
    local function getDirsFromNames(names)
        local dirs = {}
        for iName,name in ipairs(names) do
            local dirsName = videoName2batchDirs[name]
            for iDir,dir in ipairs(dirsName) do
                table.insert(dirs,dir)
            end
        end
        return dirs
    end

    math.randomseed(opt.seed) 
    if opt.overfit_batches == 0 then
        local trainsetSize = math.min(opt.trainset_size,#videoNames)

        local name2weight = {}
        for iVideo,videoName in ipairs(videoNames) do
            name2weight[videoName] = math.random()
        end
        table.sort(videoNames,function(a,b) return name2weight[a]<name2weight[b] end)
        trainsetNames={}
        validsetNames={}
        for iVideo,videoName in ipairs(videoNames) do
            if(iVideo<=trainsetSize) then
                table.insert(trainsetNames,videoName)
            else
                table.insert(validsetNames,videoName)
            end
        end
        print(string.format('%d for trainset and %d for validset',#trainsetNames,#validsetNames))

        trainsetDirs = getDirsFromNames(trainsetNames)
        validsetDirs = getDirsFromNames(validsetNames)
        table.sort(trainsetDirs)
        table.sort(validsetDirs)
        print(string.format('Split batches into %d for trainset and %d for validset',#trainsetDirs,#validsetDirs))
    else
        print(string.format( "Overfitting %d batches",opt.overfit_batches ))
        local datasetDirs = getDirsFromNames(videoNames)
        for iBatch = 1,opt.overfit_batches do
            table.insert(trainsetDirs,datasetDirs[iBatch])
        end
    end

elseif #opt.data_trainset > 0 and #opt.data_validset > 0 then

    trainsetFolder = opt.data_trainset
    print(string.format('Scanning trainset batches from %s',trainsetFolder))
    for batchDir in lfs.dir(trainsetFolder) do
        if batchDir ~= "." and batchDir ~= ".." then
            table.insert(trainsetDirs,paths.concat(trainsetFolder,batchDir));
        end
    end
    table.sort(trainsetDirs)
    print(string.format('Found %d trainset batches',#trainsetDirs))

    validsetFolder = opt.data_validset
    print(string.format('Scanning validset batches from %s',validsetFolder))
    
    for batchDir in lfs.dir(validsetFolder) do
        if batchDir ~= "." and batchDir ~= ".." then
            table.insert(validsetDirs,paths.concat(validsetFolder,batchDir));
        end
    end
    table.sort(validsetDirs)
    print(string.format('Found %d validset batches',#validsetDirs))

    if opt.overfit_batches > 0 then
        print(string.format( "Overfitting %d batches",opt.overfit_batches ))
        trainsetDirsTemp = {}
        for iBatch = 1,opt.overfit_batches do
            table.insert(trainsetDirsTemp,trainsetDirs[iBatch])
        end
        trainsetDirs = trainsetDirsTemp
    end
end

local eampleBatchNum = math.min(10,#trainsetDirs)
print(string.format( "First %d batches example of %d trainset Batch:",eampleBatchNum,#trainsetDirs ))
for iBatch = 1,eampleBatchNum do
    print('\t'..trainsetDirs[iBatch])
end
local eampleBatchNum = math.min(10,#validsetDirs)
print(string.format( "First %d batches example of %d validset Batch:",eampleBatchNum,#validsetDirs ))
for iBatch = 1,eampleBatchNum do
    print('\t'..validsetDirs[iBatch])
end


-- preload trainset if it's small
trainsetBatches = {}
if #trainsetDirs <= 100 then
    matio = require 'matio'
    print('Preloading '..#trainsetDirs..' trainset into memory')
    for iBatch,trainsetDir in ipairs(trainsetDirs) do
        table.insert(trainsetBatches, matio.load(trainsetDir))
        print(iBatch..'/'..#trainsetDirs)
    end
end


-- load model
require 'cunn'
require 'cutorch'
require 'nn'
require 'optim'

modelDir = opt.model
print(string.format( "Loading model %s",modelDir ))
require(modelDir)
net = create_model()
net = net:cuda()
print("Model loaded")


-- params setting
max_intensity = opt.max_intensity
batchSize = opt.batch_size
decayFrom = opt.decay_from
decayEvery = opt.decay_every
decayRate = opt.decay_rate
lrMin = 1e-6
itMax = opt.it_max

print(string.format( "Params setting: max_intensity = %f",max_intensity ))
print(string.format( "                    batchSize = %d",batchSize ))
print(string.format( "                    decayFrom = %d",decayFrom ))
print(string.format( "                   decayEvery = %d",decayEvery ))
print(string.format( "                    decayRate = %f",decayRate ))
print(string.format( "                        itMax = %d",itMax ))
print(string.format( "                        lrMin = %f",lrMin ))

criterion = nn.MSECriterion()
criterion = criterion:cuda()

params, gradParams = net:getParameters()

paramsSaveDir = opt.model_param
bn_meanvarSaveDir = opt.bn_meanstd
optimstateSaveDir = opt.optimstate
saveEvery = opt.save_every


-- load params
paramsLoadDir = opt.model_param_load
bn_meanvarLoadDir = opt.bn_meanstd_load
optimstateLoadDir = opt.optimstate_load

function loadParams()
    print(string.format(
    "Loading params from: \nparamsLoadDir: %s \nbn_meanvarLoadDir: %s \noptimstateLoadDir: %s ", 
    paramsLoadDir,bn_meanvarLoadDir,optimstateLoadDir))
    if not pcall(
            function()
                local paramsSave = torch.load(paramsLoadDir)
                local bn_meanvarSave = torch.load(bn_meanvarLoadDir)
                if opt.reset_state==0 then
                    optimstateSave = torch.load(optimstateLoadDir)
                end
                assert(params:nElement() == paramsSave:nElement(), string.format('%s: %d vs %d', 'loading parameters: dimension mismatch.', params:nElement(), paramsSave:nElement()))
                params:copy(paramsSave)
                local bn_mean, bn_std = table.unpack(bn_meanvarSave)
                for k,v in pairs(net:findModules('nn.SpatialBatchNormalization')) do
                    v.running_mean:copy(bn_mean[k])
                    v.running_var:copy(bn_std[k])
                end
            end) 
    then
        print('No params to load or wrong dir!\n')
    end
    
    optimConfig = optimstateSave or {}
    optimConfig.learningRate = optimConfig.learningRate or 0.005 --default lr=0.005
    if opt.reset_lr > 0 then
        optimConfig.learningRate = opt.reset_lr
    end
    optimConfig.weightDecay = optimConfig.weightDecay or 0
    optimConfig.beta1 = optimConfig.beta1 or 0.9
    optimConfig.beta2 = optimConfig.beta2 or 0.999
    optimConfig.epsilon = optimConfig.epsilon or 1e-8
    print('optimConfig = ')
    print(optimConfig)
    optimConfig.t = optimConfig.t or 0
    itBegin = optimConfig.t+1
end

loadParams()


-- load random batch sample
function copyBatch(batch)
    local batchCopy = {}
    batchCopy.batchInputTorch = batch.batchInputTorch:clone()
    batchCopy.batchGTTorch = batch.batchGTTorch:clone()
    return batchCopy
end
matio = require 'matio'
function loadRandomBatchFrom(batchDirs,batchSize)
    local iBatch = math.random(#batchDirs)
    local batchSample = nil
    if type(batchDirs[1]) == 'string' then
        local sampleDir = batchDirs[iBatch]
        batchSample = matio.load(sampleDir)
    else
        batchSample = copyBatch(batchDirs[iBatch])
    end

    local batchInputRaw,batchGTRaw = batchSample.batchInputTorch, batchSample.batchGTTorch

    batchSize = batchSize or batchInputRaw:size(1)
    if batchSize - batchInputRaw:size(1) ~= 0 then
        local shuffle = torch.randperm(batchInputRaw:size(1))
        batchInput = torch.zeros(batchSize,batchInputRaw:size(2),batchInputRaw:size(3),batchInputRaw:size(4))
        batchGT = torch.zeros(batchSize,batchGTRaw:size(2),batchGTRaw:size(3),batchGTRaw:size(4))
        for i=1,batchSize do
            batchInput[i] = batchInputRaw[shuffle[i] ]
            batchGT[i] = batchGTRaw[shuffle[i] ]
        end
    else
        batchInput = batchInputRaw
        batchGT = batchGTRaw
    end

    return batchInput, batchGT
end


-- save params
local function saveParams(it)
    if paramsSaveDir ~= '' and bn_meanvarSaveDir ~= '' and optimstateSaveDir ~= '' then
        local paramsSaveDir = paramsSaveDir..'_iteration_'..it..'.t7'
        local bn_meanvarSaveDir = bn_meanvarSaveDir..'_iteration_'..it..'.t7'
        local optimstateSaveDir = optimstateSaveDir..'_iteration_'..it..'.t7'
        print(string.format(
            "Saving params to: \nparamsSaveDir: %s \nbn_meanvarSaveDir: %s \noptimstateSaveDir: %s ", 
            paramsSaveDir,bn_meanvarSaveDir,optimstateSaveDir))
        torch.save(paramsSaveDir,params)
        local bn_mean = {}
        local bn_std = {}
        for k,v in pairs(net:findModules('nn.SpatialBatchNormalization')) do
            table.insert(bn_mean,v.running_mean)
            table.insert(bn_std,v.running_var)
        end
        local bn_meanvar = {bn_mean,bn_std}
        torch.save(bn_meanvarSaveDir,bn_meanvar)
        torch.save(optimstateSaveDir,optimConfig)
        print(optimConfig)
    else
        print('No dir to save!\n')
    end
end


-- preload batch if overfit_batches = 1
if opt.overfit_batches == 1 then
    batchInput, batchGT = loadRandomBatchFrom(trainsetDirs,batchSize)
    if opt.overfit_batches == 1 and opt.overfit_patches > 0 then
        batchInput = batchInput[{{1,opt.overfit_patches},{},{},{}}]
        batchGT = batchGT[{{1,opt.overfit_patches},{},{},{}}]
    end
    batchInput = batchInput:double():div(max_intensity):cuda()
    batchGT = batchGT:double():div(max_intensity):cuda()
end


-- iteration funtion
local function feval(params)
    gradParams:zero()
    if #trainsetBatches > 0 then
        batchInput, batchGT = loadRandomBatchFrom(trainsetBatches,batchSize)
    else
        batchInput, batchGT = loadRandomBatchFrom(trainsetDirs,batchSize)
    end
    batchInput = batchInput:double():div(max_intensity):cuda()
    batchGT = batchGT:double():div(max_intensity):cuda()

    local batchOutput = net:forward(batchInput)
    loss = criterion:forward(batchOutput, batchGT)
    local dloss_dbatchOutput = criterion:backward(batchOutput,batchGT)
    net:backward(batchInput, dloss_dbatchOutput)
        
    return loss, gradParams
end


-- train
-- Log results to files
if opt.log ~= '' then
    paths.mkdir(string.match(opt.log, "(.+)/[^/]*%.%w+$"))
    print('logging to: '..opt.log)
    trainLogger = optim.Logger(opt.log)
    trainLogger:setNames{'loss'}
end
-- at least 12 GB video memory required for batch size 64
math.randomseed(sys.clock()) 
net:training() -- set train = true
local tic = sys.clock()
lossAvg = 0;
for it = itBegin,itMax do
    if it>=decayFrom and (it-decayFrom)%decayEvery==0 then
        optimConfig.learningRate = optimConfig.learningRate*decayRate
    end
    if optimConfig.learningRate < lrMin then
        optimConfig.learningRate = lrMin
    end

    optim.adam(feval,params,optimConfig)
    
    local toc = sys.clock()
    speed = 1/(toc-tic)
    avgSpeed = avgSpeed or speed
    avgSpeed = 0.1*speed+0.9*avgSpeed
    timeLeft = (itMax-it)/avgSpeed
    toc = tic
    tic = sys.clock()
    timeLeftMin = timeLeft/60
    print(string.format(
            'it: %d, loss: %f, lr: %f, spd: %.2f it/s, avgSpd: %.2f it/s, left: %d h %.1f min',
            it,loss,optimConfig.learningRate,speed,avgSpeed,timeLeftMin/60,timeLeftMin%60))
    lossAvg = lossAvg + loss
    -- update logger/plot
    if trainLogger and it%opt.log_every == 0 then
        lossAvg = lossAvg/opt.log_every
        trainLogger:add{lossAvg}
        trainLogger:style{'-'}
        trainLogger:plot()
        lossAvg = 0
    end

    if it%opt.save_every == 0 or it == itMax then 
        saveParams(it)
    end
end


-- test on trainset for overfitting
require 'image'
if opt.overfit_batches > 0 then
    print('Testing on trainset for overfitting')
    local outImgSaveDir =  opt.overfit_out
    local outputSaveDir = paths.concat(outImgSaveDir,'output')
    local gtSaveDir = paths.concat(outImgSaveDir,'GT')
    local inputSaveDir = paths.concat(outImgSaveDir,'input')
    paths.mkdir(outputSaveDir)
    paths.mkdir(gtSaveDir)
    paths.mkdir(inputSaveDir)

    print(string.format("Saving test result to: %s",outImgSaveDir))
    net:evaluate() -- get different loss?
    local lossTest = 0
    local maxLossTest = 0
    local tic = sys.clock()
    for iBatch,trainsetDir in ipairs(trainsetDirs) do
        local batchSample = nil
        if #trainsetBatches > 0 then
            batchSample = copyBatch(trainsetBatches[iBatch])
        else
            batchSample = matio.load(trainsetDir)
        end
        local batchInput,batchGT = batchSample.batchInputTorch, batchSample.batchGTTorch
        if opt.overfit_batches == 1 and opt.overfit_patches > 0 then
            batchInput = batchInput[{{1,opt.overfit_patches},{},{},{}}]
            batchGT = batchGT[{{1,opt.overfit_patches},{},{},{}}]
        end
        batchInput = batchInput:double():div(max_intensity):cuda()
        batchGT = batchGT:double():div(max_intensity):cuda()

        local batchOutput = net:forward(batchInput)
        local loss = criterion:forward(batchOutput, batchGT)
        maxLossTest = math.max(maxLossTest,loss)
        lossTest = lossTest+loss

        batchOutput:mul(max_intensity)
        batchGT:mul(max_intensity)
        batchInput = batchInput:contiguous():view(batchInput:size(1),5,3,128,128)
        batchInput:mul(max_intensity)
        for iPatch = 1,batchOutput:size(1) do
            imName = string.match(trainsetDir,".+/([^/]*)%.%w+$")
            image.save(paths.concat(outputSaveDir,string.format('%s_%02d.jpg',imName,iPatch-1)), batchOutput[iPatch]:byte())
            image.save(paths.concat(gtSaveDir,string.format('%s_%02d.jpg',imName,iPatch-1)), batchGT[iPatch]:byte())
            image.save(paths.concat(inputSaveDir,string.format('%s_%02d.jpg',imName,iPatch-1)), batchInput[iPatch][3]:byte())
        end

        local toc = sys.clock()
        speed = 1/(toc-tic)
        timeLeft = (#trainsetDirs-iBatch)/speed
        toc = tic
        tic = sys.clock()
        timeLeftMin = timeLeft/60
        print(string.format(
                'ba: %d/%d, loss: %f, spd: %.2f ba/s, left: %d h %.1f min',
                iBatch,#trainsetDirs,loss,speed,timeLeftMin/60,timeLeftMin%60))

    end
    lossTest = lossTest/#trainsetDirs
    print(string.format('Final test loss: %f, max test loss: %f',lossTest,maxLossTest))
end

File: train/train.py
import argparse

parser = argparse.ArgumentParser()

parser.add_argument('--seed', type=int, default=251, help='rand seed')

parser.add_argument('--model', type=str, default='', help='model name')
parser.add_argument('--ckp_dir_load', type=str, default='', help='checkpoint dir to load')

# parser.add_argument('--num_frames', type=int, default=5, help='number of frames in input stack')
# parser.add_argument('--num_channels', type=int, default=3, help='rgb input')
parser.add_argument('--max_intensity', type=int, default=255, help='maximum intensity of input')

# parser.add_argument('--data_root', type=str, default='', help='folder for datasets')
parser.add_argument('--data_trainset', type=str, default='', help='folder for transet data')
parser.add_argument('--data_validset', type=str, default='', help='folder for validset data')
parser.add_argument('--batch_size', type=int, default=64,
                    help='size of batch sampled from batch files (if is larger than size from batch files, use entire batch every iteration)')
parser.add_argument('--it_max', type=int, default=80000, help='max number of iterations')

parser.add_argument('--ckp_dir', type=str, default='', help='checkpoint dir to save')
parser.add_argument('--save_every', type=int, default=500, help='auto save every save_every iterations')
parser.add_argument('--log', type=str, default='', help='dir to save log')
parser.add_argument('--log_every', type=int, default=1, help='log every log_every iterations')

parser.add_argument('--reset_lr', type=float, default=None, help='reset lr to reset_lr instead of default or saved lr')
# parser.add_argument('--reset_state', type=int, default=0, help='reset optim state')
parser.add_argument('--decay_from', type=int, default=24000, help='decay learning rate from decay_from iterations')
parser.add_argument('--decay_every', type=int, default=8000, help='decay learning rate every decay_every iteration')
parser.add_argument('--decay_rate', type=int, default=0.5, help='decay learning rate')

# parser.add_argument('--overfit_batches', type=int, default=0, help='overfit batch num for debuging')
# parser.add_argument('--overfit_patches', type=int, default=0, help='overfit patch num, valid if overfit_batches = 1')
# parser.add_argument('--overfit_out', type=str, default='', help='overfit output on trainset')

args = parser.parse_args()


# scan batches
import os


def scanSetFolder(folder):
    print('Scanning trainset batches from %s' % folder)
    dirs = os.listdir(folder)
    dirs = [os.path.join(folder, dir) for dir in dirs]
    print('Found %d trainset batches' % len(dirs))
    return dirs


def showBatchDirExamples(dirs):
    num = min(10, len(dirs))
    print("First %d batches example of %d trainset Batch:" % (num, len(dirs)))
    for i in range(num):
        print('\t' + dirs[i])


trainsetDirs = []
validsetDirs = []
if args.data_trainset != '':
    trainsetDirs = scanSetFolder(args.data_trainset)
    showBatchDirExamples(trainsetDirs)
if args.data_validset != '':
    validsetDirs = scanSetFolder(args.data_validset)
    showBatchDirExamples(validsetDirs)


# preload trainset if it's small


# load model
import tensorflow as tf
import importlib

modelDir = args.model
print("Loading model %s" % modelDir)
specImport = importlib.util.spec_from_file_location('model', modelDir)
model = importlib.util.module_from_spec(specImport)
specImport.loader.exec_module(model)

tf.reset_default_graph()
inputPh = tf.placeholder(tf.float32, shape=(None, 128, 128, 15), name='input')
gtPh = tf.placeholder(tf.float32, shape=(None, 128, 128, 3), name='gt')
trainingPh = tf.placeholder(tf.bool, shape=(), name='training')
learningRatePh = tf.placeholder(tf.float32, shape=(), name='learningRate')
learningRateT = tf.Variable(0.005,trainable=False,dtype=tf.float32)
global_step = tf.train.create_global_step()
optimizer = tf.train.AdamOptimizer(learning_rate=learningRatePh)
outputT = model.create_model(inputPh, trainingPh)
lossT = tf.losses.mean_squared_error(gtPh, outputT)

updateOps = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(updateOps):
    trainOp = optimizer.minimize(lossT, global_step=global_step)
print("Model loaded")


# params setting
maxIntensity = args.max_intensity
batchSize = args.batch_size
decayFrom = args.decay_from
decayEvery = args.decay_every
decayRate = args.decay_rate
lrMin = 1e-6
itMax = args.it_max

print("Params setting: max_intensity = %f" % maxIntensity)
print("                    batchSize = %d" % batchSize)
print("                    decayFrom = %d" % decayFrom)
print("                   decayEvery = %d" % decayEvery)
print("                    decayRate = %f" % decayRate)
print("                        itMax = %d" % itMax)
print("                        lrMin = %f" % lrMin)

# load params
gpu_options = tf.GPUOptions(allow_growth=True)
sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options))
saver = tf.train.Saver()
if args.ckp_dir_load != '':
    loadDir = os.path.join(args.ckp_dir_load, 'checkpoints')
    loadDir = tf.train.latest_checkpoint(args.ckp_dir_load)
    try:
        print('Loading params from: ' + loadDir)
        saver.restore(sess, loadDir)
    except:
        print('Load params failed!')
        sess.run(tf.global_variables_initializer())
else:
    sess.run(tf.global_variables_initializer())

if args.reset_lr != None:
    learningRateT.load(args.reset_lr, sess)

print('Learning rate = %f' % sess.run(learningRateT))


# load random batch sample
from random import choice
import scipy.io as sio
import numpy as np

def loadRandomBatchFrom(batchDirs, batchSize=None):
    sampleDir = choice(batchDirs)
    batchSample = sio.loadmat(sampleDir)
    batchInputRaw = batchSample['batchInputTorch']
    batchGTRaw = batchSample['batchGTTorch']

    if batchSize is None:
        batchSize = batchInputRaw.shape[0]
    if batchSize == batchInputRaw.shape[0]:
        batchInput = batchInputRaw
        batchGT = batchGTRaw
    else:
        idxs = np.random.choice(batchInputRaw.shape[0], batchSize, replace=False)
        print(idxs)
        batchInput = batchInputRaw[idxs, :, :, :]
        batchGT = batchGTRaw[idxs, :, :, :]
    # change to NHWC
    batchInput = batchInput.transpose((0, 2, 3, 1)).astype(np.float32) / maxIntensity
    batchGT = batchGT.transpose((0, 2, 3, 1)).astype(np.float32) / maxIntensity

    return batchInput, batchGT


# preload batch if overfit_batches = 1



# train
import time
if args.log != '':
    lossAvg = 0
    writer = tf.summary.FileWriter(args.log, sess.graph)
    lossAvgPh = tf.placeholder(dtype=tf.float32)
    tf.summary.scalar('loss', lossAvgPh)
    tf.summary.scalar('learning rate', learningRateT)
    tf.summary.image('inputIm',tf.slice(inputPh, [0, 0, 0, 6], [-1, -1, -1, 3]))
    tf.summary.image('outputIm', outputT)
    merged = tf.summary.merge_all()
    print('Logging to: ' + args.log)

tic = time.time()
it = tf.train.global_step(sess, global_step)
avgSpeed = None
while 1:
    if it >= itMax:
        break
    if it >= decayFrom and (it - decayFrom) % decayEvery == 0:
        sess.run(learningRateT.assign(learningRateT * decayRate))
    if learningRateT.eval(sess) < lrMin:
        learningRateT.load(lrMin, sess)
    batchInput, batchGT = loadRandomBatchFrom(trainsetDirs)
    batchLoss, _ = sess.run(
        (lossT, trainOp),
        feed_dict={
            learningRatePh: learningRateT.eval(sess),
            trainingPh: True,
            inputPh: batchInput,
            gtPh: batchGT
        }
    )
    it = tf.train.global_step(sess, global_step)

    toc = time.time()
    speed = 1 / (toc - tic)
    avgSpeed = avgSpeed or speed
    avgSpeed = 0.1 * speed + 0.9 * avgSpeed
    timeLeft = (itMax - it) / avgSpeed
    toc = tic
    tic = time.time()
    timeLeftMin = timeLeft / 60
    print('it: %d, loss: %f, lr: %f, spd: %.2f it/s, avgSpd: %.2f it/s, left: %d h %.1f min' %
          (it, batchLoss, learningRateT.eval(sess), speed, avgSpeed, timeLeftMin / 60, timeLeftMin % 60))
    lossAvg = lossAvg + batchLoss
    if args.log != '' and it % args.log_every == 0:
        lossAvg = lossAvg/args.log_every
        rs = sess.run(
            merged,
            feed_dict={
                learningRatePh: learningRateT.eval(sess),
                trainingPh: True,
                inputPh: batchInput,
                gtPh: batchGT,
                lossAvgPh:lossAvg
            })
        writer.add_summary(rs, it)
        print('Logging to: ' + args.log)
        lossAvg = 0

    if it % args.save_every == 0 or it == itMax:
        if args.ckp_dir != '':
            if not os.path.exists(args.ckp_dir):
                os.makedirs(args.ckp_dir)
            savePath = saver.save(
                sess,
                os.path.join(args.ckp_dir, 'checkpoints'),
                global_step=it
            )
            print("Model saved in path: %s" % savePath)
        else:
            print('Use param --ckp_dir to specify path to save!')

writer.close()


# test net
import matplotlib.pyplot as plt

def showBatchIm(batch, i):
    im = batch[i, :, :, :]
    plt.imshow(im)
    plt.show()

batchInput, batchGT = loadRandomBatchFrom(trainsetDirs, 1)
inputIm = batchInput[:, :, :, 6:9]
showBatchIm(inputIm, 0)
showBatchIm(batchGT, 0)

def testNet(batchInput, batchGT, sess):
    batchOutput, batchLoss = sess.run(
        (outputT, lossT),
        feed_dict={
            trainingPh: False,
            inputPh: batchInput,
            gtPh: batchGT
        }
    )
    print('loss = %f' % batchLoss)
    showBatchIm(batchOutput, 0)


testNet(batchInput, batchGT, sess)

# test on trainset for overfitting
